Add tensor parallelism on QLoRA

tugang-baidu opened the pull request:
# Updates
In QLoRA, the weight quantization algorithm is nf4, which is not supported in neither ColumnParallelQuantizationLinear nor RowParallelQuantizationLinear and thus cannot be run with tensor parallelism. Following updates can support running QLoRA with tensor parallelism.

In paddleformers/quantization/quantization_linear.py：copy codes about handling nf4 and fp4 in `class QuantizationLinear` to both `ColumnParallelQuantizationLinear` and `RowParallelQuantizationLinear` and do some changes in reference to codes about handling weight_only_int8.

In paddleformers/transformers/conversion_utils.py: in method `add_quant_mapping`, add nf4 to the list shown in `post_quantize = quantization_config ... `and `post_quantize = any ...`, otherwise tensor parallelism cannot find the corresponding names in layer and thus will fail.

Besides, using nf4 will lead to a operator error:
`ValueError: (InvalidArgument) The type of data we are trying to retrieve (float32) does not match the type of data (bfloat16) currently contained in the container.`

Following updates can fix it:

In paddleformers/peft/lora/init.py: add `from .lora_quantization_layers import QuantizationLoRABaseLinear` so that we can import QuanzationLoRABaseLinear.

In paddleformers/trainer/trainer.py: add `from ..peft.lora import QuantizationLoRABaseLinear` and add QuantizationLoRABaseLinear in `excluded_layers` of `paddle.amp.decorate`


In paddleformers/transformers/model_utils.py: add "nf4" and "fp4" in the list of the check of post_quantize in Line 2048

By the way, there is a spelling error in paddleformers/quantization/quantization_utils.py:
quant_sacle_offset_name = name + ".quant_sacle_offset"
the word "sacle" in "quant_sacle_offset" should be spelled as "scale".
However, once it's changed, you need to change all 'quant_sacle_offset' in the project. It's recommended to rewrite it as 'weight_scale_offset' to keep alignment with that name in 'quantization_linear.py'.

# Experimental Results
I test tensor_parallel_degree = 1 and tensor_parallel_degree = 2 on ERNIE-4.5-0.3B.

The two curves of train loss mostly overlap:
<img width="2736" height="1338" alt="image" src="https://github.com/user-attachments/assets/b8f7e050-bf2f-4168-8dad-dc9336a110c7" />

And I test tensor_parallel_degree = 2, pipeline_parallel_degree = 4 on ERNIE-4.5-300B-A47B:
<img width="1926" height="956" alt="image" src="https://github.com/user-attachments/assets/0194012d-cf7a-40af-832d-76d7dee13513" />

Results of another comparision experiments on ERNIE-4.5-21B-A3B are as follows, green is for QLoRA, blue is for LoRA with quantization of WINT4/8 and purple is for LoRA. The three curves mostly overlap.
<img width="3794" height="2246" alt="image" src="https://github.com/user-attachments/assets/4f0223d3-1387-4420-b6f4-c3a6e517eb13" />

tugang-baidu committed:
Update lora_quantization_layers.py to fix parallel QLoRA

Fix parallel qlora in reference to quantization_linear.py

In __init__() of class QuantizationLoRABaseLinear:

Insert codes between "self.bias = layer.bias" and "self.lora_config = lora_config":

        self.state = 0
        if self.weight_quantize_algo in ["a8w8linear", "a8w4linear", "fp8linear"]:
            self.act_scale = self.create_parameter(
                shape=[1],
                dtype=self._dtype,
                is__bias=False,
                default_initializer=nn.initializer.Constant(value=0.0),
            )
            self.act_scale.is_distributed = False
            self.act_scale.stop_gradient = True
            self.group = get_act_scale_group(is_row=True)
        else:
            raise NotImplementedError(f"Not supported weight_quantize_algo {self.weight_quantize_algo}")

In forward, define act_scale = (self.state, self.training, self.act_scale, self.group) and put it in the parameter list of "output = quant_weight_linear"

Insert codes before "return output":
        if self.training:
            self.state += 1

tugang-baidu committed:
Update lora_quantization_layers.py

change 'is__bias' to 'is_bias'

tugang-baidu committed:
Update lora_quantization_layers.py

delete duplicate tabs of 'self.state=0'

tugang-baidu committed:
Update quantization_linear.py

Add actions of partioning quant_weight, weight_scale, etc. of weight_quantize_algo "nf4" and "fp4" in both ColumnParallelQuantizationLinear and RowQuantizationLinear

tugang-baidu committed:
Update quantization_utils.py

Correct the spelling error and change the word "sacle" to "scale"

tugang-baidu committed:
Update conversion_utils.py

In method add_quant_mapping, add "nf4" to the list in post_quantize = quantization_config .... and post_quantize = any ... otherwise a debug message like " tensor parallel conversion: could not find name layers.10.self_attn. qkv_proj. quant_weight in loaded state dict! " will show on the console which directly leads to failure of tensor parallelism.

tugang-baidu committed:
Update trainer.py

add QuantizationLoRABaseLinear to excluded_layers in paddle.amp.decorate to avoid _ops_ error

tugang-baidu committed:
Update __init__.py

add from .lora_quantization_layers import QuantizationLoRABaseLinear

tugang-baidu committed:
Update model_utils.py

add nf4 and fp4 in the list of check of post_quantize to avoid operator error

lugimzzz commented on paddleformers/peft/lora/lora_quantization_layers.py:
```diff
@@ -44,6 +44,21 @@ def __init__(self, layer, lora_config):
         else:
             self.weight_scale = layer.weight_scale
         self.bias = layer.bias
+        self.state = 0
```
不太理解为什么新增qlora要加这段代码

tugang-baidu replied:
Sorry for that. In paddlenlp, when you use 'weight_only_int8' as weight quantization algorithm, in the `forward `method of `class QuantizationLoRABaseLinear`, you'll find that it doesn't give `act_scale` which will be set to None by default and thus incur error showing `Cannot Unpack Non-iterable Nonetype Objects in Python`. In paddleformers, using weight_only_int8 works normally without extra variables and giving `act_scale`.

DesmonDay replied:
这个PR下只需要考虑 paddleformers

tugang-baidu replied:
By the way, here `self.state=0` in paddlenlp should be rewritten as `self.state=layer.state`. I have deleted this and `act_scale = ...` and `if self.training:...` below.

lugimzzz commented on paddleformers/quantization/quantization_linear.py:
```diff
@@ -496,6 +496,67 @@ def __init__(
+                self.qweight_scale.is_distributed = True if self.is_mp else False
+                if self.qweight_scale.is_distributed:
+                    self.qweight_scale.split_axis = 0
+                self.double_weight_scale = self.create_parameter(
```
试一下double_quant分支打开是否有问题

tugang-baidu replied:
New experiments will be conducted soon.

tugang-baidu replied:
With tensor_parallel_degree = 2 and qlora_weight_double_quant = True, the experiment result on ERNIE-4.5-0.3B is:
<img width="2360" height="1364" alt="659ff6d16f1c97f4e645e15b23d08335" src="https://github.com/user-attachments/assets/52667bfb-f59d-47f8-bb58-adb08c7f1d5c" />

tugang-baidu committed:
Update lora_quantization_layers.py

Delete unnecessary codes.

tugang-baidu committed:
Update qlora.py

Change quant_sacle_offset_name from 'quant_sacle_offset' to 'weight_scale_offset' to keep alignment with class QuantizationLinear.

tugang-baidu committed:
Update quantization_utils.py

change 'quant_scale_offset' and 'quant_sacle_offset' to 'weight_scale_offset' to keep alignment with class QuantizationLinear and  method qlora_weight_quantize in qlora.py

tugang-baidu committed:
Update quantization_linear.py

In ColumnParallelQuantizationLinear:
set self.weight_scale_offset.is_distributed = True if self.is_mp else False and self.weight_scale_offset.split_axis = 0 if self.weight_scale_offset.is_distributed.

In RowParallelQuantizationLinear:
1. change self.qweight_scale.split_axis from 1 to 0.
2. set self.weight_scale_offset.is_distributed = True if self.is_mp else False and self.weight_scale_offset.split_axis = 0 if self.weight_scale_offset.is_distributed.

lugimzzz commented:
解决一下codestyle问题，提交代码前先pre-commit install

lugimzzz commented on paddleformers/peft/lora/lora_quantization_layers.py:
```diff
@@ -45,6 +45,21 @@ def __init__(self, layer, lora_config):
             self.weight_scale = layer.weight_scale
         self.bias = layer.bias
 
+        if self.weight_quantize_algo in ["a8w8linear", "a8w4linear", "fp8linear"]:
```
为什么要修改这里？

lugimzzz replied:
不应该在这里create parameters

tugang-baidu replied:
These unnecessary codes have been deleted in my fork.

tugang-baidu committed:
Update lora_quantization_layers.py

Delete unnecessary codes self.act_scale = ...

tugang-baidu committed:
Update lora_quantization_layers.py

delete unused import of get_act_scale_group.

tugang-baidu committed:
Create pre-commit.yaml

tugang-baidu committed:
Update model_utils.py

Codestyle changes.

tugang-baidu committed:
Update conversion_utils.py

Some codestyle changes.

tugang-baidu committed:
Update model_utils.py

Some codestyle changes.

tugang-baidu committed:
Update tokenizer_utils.py

Some codestyle changes.

tugang-baidu committed:
Update lora_quantization_layers.py

Some codestyle changes.

tugang-baidu committed:
Update quantization_linear.py

Some codestyle changes.

tugang-baidu committed:
Update pre-commit.yaml

Update flake8

tugang-baidu committed:
Update trainer.py

Some codestyle changes

tugang-baidu committed:
Update pre-commit.yaml

tugang-baidu committed:
Update and rename pre-commit.yaml to pre-commit-config.yaml

tugang-baidu committed:
Delete .github/workflows/pre-commit-config.yaml

tugang-baidu committed:
Create pre-commit-config.yaml

tugang-baidu committed:
Update .pre-commit-config.yaml

tugang-baidu committed:
Delete pre-commit-config.yaml

tugang-baidu committed:
Update .pre-commit-config.yaml

tugang-baidu committed:
Update quantization_linear.py

Apply Black formatting fixes

tugang-baidu committed:
Update conversion_utils.py

Apply Black formatting fixes

tugang-baidu committed:
Update model_utils.py

Apply Black formatting fixes

tugang-baidu committed:
Update trainer.py

Apply Black formatting fixes

tugang-baidu committed:
Update lint.yml

Free runner disk space

tugang-baidu committed:
Update lint.yml

tugang-baidu committed:
codestyle changes

tugang-baidu committed:
recover

tugang-baidu committed:
update lint.yml

tugang-baidu committed:
update lint.yml

tugang-baidu committed:
update lint.yml

tugang-baidu committed:
update lint.yml

tugang-baidu committed:
recover lint.yml

tugang-baidu committed:
Trigger rerunning unitest-gpu-ci

lugimzzz reviewed (approved):
LGTM
